frontend/src/metabase/query_builder/utils.ts

```diff
@@ -79,23 +79,45 @@ export const isNavigationAllowed = ({
   }
 
   const { hash, pathname } = destination;
+  const isRunningModel = pathname === "/model" && hash.length > 0;
+  const isRunningQuestion = pathname === "/question" && hash.length > 0;
+  const validSlugs = [question.id(), question.slug()]
+    .filter(Boolean)
+    .map(String);
 
   if (question.isDataset()) {
     if (isNewQuestion) {
-      const isQueryTab = pathname === "/model/query";
-      const isMetadataTab = pathname === "/model/metadata";
-      return isQueryTab || isMetadataTab;
+      const allowedPathnames = ["/model/query", "/model/metadata"];
+      return isRunningModel || allowedPathnames.includes(pathname);
     }
 
-    const isQueryTab = pathname.match(/^\/model\/.+\/query$/) !== null;
-    const isMetadataTab = pathname.match(/^\/model\/.+\/metadata$/) !== null;
-    return isQueryTab || isMetadataTab;
+    const allowedPathnames = validSlugs.flatMap(slug => [
+      `/model/${slug}`,
+      `/model/${slug}/query`,
+      `/model/${slug}/metadata`,
+    ]);
```

Comment:
If user opens `/question/4/notebook`, modifies & saves it, they will be redirected to `/question/4-some-slug/notebook`. We need to support both (some tests fail otherwise).

BTW I managed to [avoid the regex](https://github.com/metabase/metabase/pull/34527#discussion_r1356961607) after all. cc: @uladzimirdev :)